_posts/2021-05-28-why-avoid-an-orm.md

```diff
@@ -0,0 +1,231 @@
+ 
+As far as I know, ORMs tend to handle this by forbidding object identity (don’t override object equality) and always use the database identity. Too bad you now have to live with objects that look equal but are not. A company named "Solita" isn’t equal to a company named "Solita” until you fetch them both from the database to compare their keys.
+ 
+What if you want to fetch objects that don't have an identity? This is quite common in a relational database. The ORM would have to support result objects _with_ identity and result objects _without_ identity. [Hibernate](https://hibernate.org) or _Java Persistence API_ doesn't, so apparently, it's not trivial to support for some reason. If you do `SELECT firstname, lastname FROM employees` you don't have any identifiers and easily get duplicate rows, which is fine until your ORM forces you to represent everything as objects with identity.
+ 
+When you want to change the department of an employee, in the object model you'd probably say something like this:
+```java
+void updateDepartment(Id<Employee> empId, Id<Department> newDepId) {
+  Employee emp = em.get(empId);
+  emp.setDepartment(newDepId); // uh oh
+}
+```
+but you can't since the object model expects actual departments, not just some identifiers that repsesent them. So you have to go and fetch the department from the database even though you don't actually need it:
+```java
+void updateDepartment(Id<Employee> empId, Id<Department> newDepId) {
+  Employee emp = em.get(empId);
+  Department dep = em.get(newDepId);
+  emp.setDepartment(dep);
+}
+```
+ 
+To handle this performance problem, some ORMs provide a way to fetch "a department which isn't actually a department but just looks like one" which doesn't need to go to the database when you already have its identifier. Now we are again in the rabbit hole of proxies since the department is now an instance of a proxy class that will work like a regular entity. To do that, it will fetch the data from the database when you call any of its methods. Thus we face again the problem where an innocent `department.getName()` method call might be making a database query, and with some imaginative eager fetch specifications, might even fetch half the database to memory.
+ 
+There are [different kinds of proxying mechanisms](https://bravenewgeek.com/proxies-why-theyre-useful-and-how-theyre-implemented/). In one of them, the department object isn't an actual instance of a subclass of the `Department` class, but it's an instance of a proxy class inheriting from the `Department` class. Otherwise you can't see the difference, but when you use _inheritance_ (and you _would_ use inheritance since you are dealing with object-orientation) and you know from context that your department happens to be an instance of a `ProfitableDepartment` instead of a  `ManagementDepartment` ;)
```

Comment:
Perhaps a warning against inheritance here would make for a stronger argument, instead of arguing that it's going to be used anyway?

Reply:
Thanks, that's a good suggestion! Do you think it's better now?

Reply:
Yeah, looks good now.